torch/_inductor/codecache.py

```diff
@@ -379,6 +386,240 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
     tmp_path.rename(path)
 
 
+@dataclasses.dataclass
+class TensorMetadata:
+    """
+    The Tensor metadata relevant when hashing FxGraph cache keys.
+    """
+
+    dtype: torch.dtype
+    shape: torch.Size
+    stride: Tuple[Any, ...]
+    device: torch.device
+    requires_grad: bool
+    storage_offset: int
+    memory_format: Optional[torch.memory_format]
+    is_quantized: bool
+
+
+@dataclasses.dataclass
+class TensorMetadataAndValues:
+    """
+    TensorMetadata plus the elements as a list of raw values.
+    """
+
+    tensor_metadata: TensorMetadata
+    values: List[Any]
+
+
+def extract_tensor_metadata(t: torch.Tensor) -> TensorMetadata:
+    """
+    Extract the TensorMetadata of a tensor.
+    """
+    # Note the order here matters since is_contiguous() can be True
+    # for more than one memory format.
+    memory_formats = [
+        torch.contiguous_format,
+        torch.channels_last,
+        torch.channels_last_3d,
+    ]
+
+    memory_format = None
+    for fmt in memory_formats:
+        if t.is_contiguous(memory_format=fmt):
+            memory_format = fmt
+            break
+
+    return TensorMetadata(
+        dtype=t.dtype,
+        shape=t.shape,
+        stride=t.stride(),
+        device=t.device,
+        requires_grad=t.requires_grad,
+        storage_offset=t.storage_offset(),
+        memory_format=memory_format,
+        is_quantized=t.is_quantized,
+    )
+
+
+def _ident(x: Any) -> Any:
+    return x
+
+
+def _reduce_fake_tensor(t):
+    """
+    See FxGraphCachePickler. Custom reducer to pickle FakeTensors.
+    """
+    metadata = extract_tensor_metadata(t)
+    return (_ident, (metadata,))
+
+
+def _reduce_tensor(t):
+    """
+    See FxGraphCachePickler. Custom reducer to pickle Tensors.
+    """
+    # See GraphLowering. We expect any Tensors to be small inlined constants.
+    # For these, we hash on the metadata _and_ the actual values.
+    assert len(t.shape) == 0 or (len(t.shape) == 1 and t.shape[0] <= 128)
```

Comment:
This part isn't quite right. @eellison tells me there are some tensors for which it's safe to consider only their metadata and others (inlined constants) where we should also consider the values. This 128 in the assert is totally arbitrary.